File: src/HDRFixer.Core/ColorMath/TransferFunctions.cs
namespace HDRFixer.Core.ColorMath;

public static class TransferFunctions
{
    private const double SrgbLinearThreshold = 0.04045;
    private const double SrgbLinearScale = 12.92;
    private const double SrgbGammaOffset = 0.055;
    private const double SrgbGammaBase = 1.055;
    private const double SrgbGammaExponent = 2.4;
    private const double PqM1 = 2610.0 / 16384.0;
    private const double PqM2 = 128.0 * 2523.0 / 4096.0;
    private const double PqC1 = 3424.0 / 4096.0;
    private const double PqC2 = 32.0 * 2413.0 / 4096.0;
    private const double PqC3 = 32.0 * 2392.0 / 4096.0;

    public static double SrgbEotf(double v)
    {
        if (v <= SrgbLinearThreshold) return v / SrgbLinearScale;
        return Math.Pow((v + SrgbGammaOffset) / SrgbGammaBase, SrgbGammaExponent);
    }

    public static double SrgbInvEotf(double l)
    {
        if (l <= 0.0031308) return l * SrgbLinearScale;
        return SrgbGammaBase * Math.Pow(l, 1.0 / SrgbGammaExponent) - SrgbGammaOffset;
    }

    public static double GammaEotf(double v, double gamma) => Math.Pow(v, gamma);
    public static double GammaInvEotf(double l, double gamma) => Math.Pow(l, 1.0 / gamma);

    public static double PqEotf(double v)
    {
        double vp = Math.Pow(v, 1.0 / PqM2);
        double num = Math.Max(vp - PqC1, 0.0);
        double den = PqC2 - PqC3 * vp;
        return 10000.0 * Math.Pow(num / den, 1.0 / PqM1);
    }

    public static double PqInvEotf(double nits)
    {
        double y = Math.Pow(nits / 10000.0, PqM1);
        return Math.Pow((PqC1 + PqC2 * y) / (1.0 + PqC3 * y), PqM2);
    }
}

File: src/core/color/transfer_functions.cpp
#include "transfer_functions.h"

namespace hdrfixer::color {

double srgb_eotf(double v) {
    v = std::clamp(v, 0.0, 1.0);
    if (v <= kSrgbLinearThreshold)
        return v / kSrgbLinearScale;
    return std::pow((v + kSrgbGammaOffset) / kSrgbGammaBase, kSrgbGammaExponent);
}

double srgb_inv_eotf(double l) {
    l = std::max(l, 0.0);
    if (l <= kSrgbInvLinearThreshold)
        return l * kSrgbLinearScale;
    return std::clamp(kSrgbGammaBase * std::pow(l, 1.0 / kSrgbGammaExponent) - kSrgbGammaOffset, 0.0, 1.0);
}

double pq_eotf(double v) {
    v = std::clamp(v, 0.0, 1.0);
    double vp = std::pow(v, 1.0 / kPqM2);
    double num = std::max(vp - kPqC1, 0.0);
    double den = kPqC2 - kPqC3 * vp;
    if (den <= 0.0) return 0.0;
    return kPqMaxNits * std::pow(num / den, 1.0 / kPqM1);
}

double pq_inv_eotf(double nits) {
    nits = std::clamp(nits, 0.0, kPqMaxNits);
    double y = std::pow(nits / kPqMaxNits, kPqM1);
    return std::pow((kPqC1 + kPqC2 * y) / (1.0 + kPqC3 * y), kPqM2);
}

double gamma_eotf(double v, double gamma) {
    if (v < 0.0) return 0.0;
    return std::pow(v, gamma);
}

double gamma_inv_eotf(double l, double gamma) {
    if (l < 0.0) return 0.0;
    return std::pow(l, 1.0 / gamma);
}

} // namespace hdrfixer::color
